Menu Items overview - Filter on Component - v2

pe7er opened the pull request:
Pull Request to add a **Component filter** to the "**Menus: Items**" **overview**.

It's PR #24665 re-done for the current Joomla 4.2-dev.


### Summary of Changes
This PR adds a Component filter to the menu items overview so that you can filter on specific Components.

### Before PR
![menu-before-pr2](https://user-images.githubusercontent.com/1217850/80373801-e8797500-8895-11ea-9440-eddcf5c44062.png)

### Testing Instructions
- Install the PR.
- Open the Menus: All Menu Items page (Menus > All Menu Items)
- Create a menu item of type Contact - Single Contact
- Click on "Search Tools",
- Select a component (e.g. Contacts)

### Expected result
The list of menu items should be filtered to only show menu items to views of that selected Component.

### After PR
A new Filter field should be available:
![menu-after-pr](https://user-images.githubusercontent.com/1217850/80373598-946e9080-8895-11ea-9e38-464baa80bd01.png)

The new filter fields enables you to filter on Menu Items of type = Articles (com_content)
![menu-after-pr-articles](https://user-images.githubusercontent.com/1217850/80373596-93d5fa00-8895-11ea-8fff-5e2a78d40dc1.png)


The new filter fields enables you to filter on Menu Items of type = Contacts (com_contact)
![menu-after-pr-contacts](https://user-images.githubusercontent.com/1217850/80373592-933d6380-8895-11ea-9ca1-9ca92bee76f3.png)

pe7er committed:
Menu Items overview - Filter on Component - v2

Quy commented on administrator/components/com_menus/forms/filter_items.xml:
```diff
@@ -76,6 +76,14 @@
+			type="componentsCategory"
+			label="COM_MENUS_FILTER_COMPONENT_LABEL"
+			onchange="this.form.submit();"
+		>
```
```suggestion
			>
```

pe7er committed:
Update administrator/components/com_menus/forms/filter_items.xml

Co-Authored-By: Quy <quy@fluxbb.org>

ChristineWk commented:
I have tested this item :white_check_mark: successfully on aaf4ce3a231b42389f9a1e83c33edb0cb0bf6875<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

jwaisner commented:
I have tested this item :white_check_mark: successfully on aaf4ce3a231b42389f9a1e83c33edb0cb0bf6875<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

jwaisner commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

alikon committed:
prepared statement

alikon commented:
@pe7er look at https://github.com/pe7er/joomla-cms/pull/1it use the prepared statement for the new where clause

pe7er committed:
Merge pull request #1 from alikon/patch-109

sql prepared statement

wilsonge commented:
Wouldn't it be more obvious to filter by menu item type than component?

sere-b commented:
I have tested this item :white_check_mark: successfully on d951e96b5eafb7e278691e939fd85fa2630037f1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
Back to pending after recent change for prepared statements. One more tester needed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

ChristineWk commented:
I have tested this item :white_check_mark: successfully on d951e96b5eafb7e278691e939fd85fa2630037f1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
Back to RTC after 2 good tests. But could cou check @wilsonge 's comment "Wouldn't it be more obvious to filter by menu item type than component?"? Seems to be a valid question to me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
@pe7er Could you check [George's comment above](https://github.com/joomla/joomla-cms/pull/28832#issuecomment-622934473 "https://github.com/joomla/joomla-cms/pull/28832#issuecomment-622934473")? I think it is a valid question. When fitering by the type of menu item we may achieve nearly the same result, because the item types are connected to the component, and it might also be easier.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

pe7er committed:
Merge branch '4.0-dev' of github.com:pe7er/joomla-cms into menu-items-overview-filter-on-component

pe7er committed:
Menu Items overview - Filter on Component - v3

pe7er commented:
@richard67 @wilsonge I have changed my PR:
- added a **new Field type** MenuItemByComponent

Before the change the dropdown would be:
![joomla-menu-filter-before](https://user-images.githubusercontent.com/1217850/81682074-e9450600-9454-11ea-9a04-a7403ff9533d.png)

After the change the dropdown looks like:
![joomla-menu-filter-3](https://user-images.githubusercontent.com/1217850/81682070-e813d900-9454-11ea-938c-10789da23e73.png)



**Note:** I've added a ```->where($db->quoteName('menu.client_id') . ' = 0')``` to the query.
Otherwise it would also show menu items from the admin menu in the dropdown:
![joomla-menu-filter](https://user-images.githubusercontent.com/1217850/81682078-e9450600-9454-11ea-9528-cc55a5f5e0d2.png)

richard67 commented:
@pe7er Should the label then still be "- Select Component -"?

pe7er commented:
@richard67 yes, I think so. Or do you have a suggestion for something better?

richard67 commented:
@pe7er Well, when you create a new menu item, it is called "Menu Item Type":

![menu-item-type](https://user-images.githubusercontent.com/7413183/81688596-f6b0bf00-9459-11ea-97c0-884ad785b588.png)

So maybe we should use the same for the new filter? "- Select Menu Item Type -" ... ok, maybe that's a bit long, but would be consistent.

brianteeman commented:
That's because it contains the views for each component. If I understand correctly this pr just lists the components not all their views

richard67 commented:
I see.

richard67 commented:
@pe7er Other question. In your 2nd screenshot, banners are missing. Is that expected?

pe7er commented:
@brianteeman Yes, that's right. I wanted to filter on View and analyzed the #__menu table to see if it's easy to filter on views: it is not. The only field that has the view, has it in a larger string of text.

@richard67 good point! Yes, the filter only shows the components for which a menu item has been created. The banner component does not have a front-end view (only modules that display the banners). 

In the 1st screenshot, the banners is present because I filtered on components table. Now only the components of created front-end menu items are in the drop down list.

richard67 commented:
> The banner component does not have a front-end view (only modules that display the banners).

Ah, sure, I should have known that.

pe7er commented:
> > The banner component does not have a front-end view (only modules that display the banners).
> Ah, sure, I should have known that.

I would love to have a front-end view for the banner component though. I've had use cases in the past: a page that list all the banners of sponsors. Because of the the missing banner view I've solved that with adding an empty article + loaded the banner module. :-)

SharkyKZ commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,97 @@
+		$options = array();
+
+		$query = $db->getQuery(true);
+		$query->select('DISTINCT ' . $db->quoteName('extensions.name'))
```
Use extension's `element` instead of `name` please. It's more reliable.

pe7er replied:
ok, thanks!

SharkyKZ commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,97 @@
+			// Extract the component name and optional section name
+			$parts     = explode('.', $categoryType);
+			$component = $parts[0];
+			$section   = (count($parts) > 1) ? $parts[1] : null;
```
Looks like this was copied from category field. It's not going to work here.

pe7er replied:
Yes, you are right. I've fixed it. Thanks @SharkyKZ !

pe7er committed:
Menu Items overview - Filter on Component - v4

pe7er commented:
I've fixed the issues in MenuItemByComponentField that @SharkyKZ mentioned.

Quy commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,92 @@
+
+namespace Joomla\Component\Menus\Administrator\Field;
+
+defined('JPATH_BASE') or die;
```
```suggestion
\defined('_JEXEC') or die;
```

See #28416

pe7er replied:
Thanks! Fixed!

pe7er committed:
Menu Items overview - Filter on Component - fix

SharkyKZ commented on administrator/components/com_menus/src/Model/ItemsModel.php:
```diff
@@ -60,6 +60,7 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
 				'parent_id', 'a.parent_id',
 				'publish_up', 'a.publish_up',
 				'publish_down', 'a.publish_down',
+				'e.name', 'componentName',
```
```suggestion
				'e.element', 'componentName',
```

SharkyKZ commented on administrator/components/com_menus/src/Model/ItemsModel.php:
```diff
@@ -523,6 +524,13 @@ protected function getListQuery()
+		// Filter on componentName
+		if ($componentName = $this->getState('filter.componentName'))
+		{
+			$query->where($db->quoteName('e.name') . ' = :component')
```
```suggestion
			$query->where($db->quoteName('e.element') . ' = :component')
```

pe7er committed:
Update administrator/components/com_menus/src/Model/ItemsModel.php

Yes. Thanks!

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

pe7er committed:
Update administrator/components/com_menus/src/Model/ItemsModel.php

Thanks!

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,92 @@
+			|| $lang->load($component, JPATH_ADMINISTRATOR . '/components/' . $component);
+
+			// If the component section string exists, let's use it
+			if ($lang->hasKey($component_section_key = strtoupper($component)))
```
Also don't need this if/else. Use `$option->text = Text::_(strtoupper($component))`.

pe7er replied:
Thanks! Fixed!

SharkyKZ replied:
Removed a little too much 😄. Language files do need to be loaded.

pe7er replied:
Oops :-) Fixed!

pe7er commented:
Thank you both for your improvements @Quy + @SharkyKZ

pe7er committed:
removed unnecessary if statement

pe7er committed:
fixed the removal of language file loading

SharkyKZ commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,81 @@
+
+			$option        = new \stdClass;
+			$option->value = $component;
+			$option->text  = Text::_(strtoupper($component))
```
```suggestion
			$option->text  = Text::_(strtoupper($component));
```

pe7er replied:
Yes, I just noticed that too. Fixed!

pe7er committed:
fixed missing semicolon

pe7er commented:
@Quy @SharkyKZ @richard67 Do I need two new tests before getting it to RTC?

richard67 commented:
@pe7er I would say yes, because changes after the previous tests were not cosmetic only. I'll see if I can find time for testing tomorrow.

richard67 commented:
I have tested this item :white_check_mark: successfully on fb902a4f179fbbc58df44cd14c480d3ee45d3ad5<br /><br />The behavior of the new filter is a bit different to other filters, but I think that's intended.

Other filters always show all possible options, regardless if there are menu items using it or not, while the new filter added with this PR only shows those components in the selection forn which there is a menu item, i.e. e.g. if you don't have any menu item for com_contact, the "Contacts" is not available in the dropdown list.

@pe7er Let me know if that's not intended, then I can change back my test result.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

brianteeman commented:
> Other filters always show all possible options, regardless if there are menu items using it or not, while the new filter added with this PR only shows those components in the selection forn which there is a menu item, i.e. e.g. if you don't have any menu item for com_contact, the "Contacts" is not available in the dropdown list.

I hope its the intended behaviour. Look at the module filter in com_modules

richard67 commented:
> I hope its the intended behaviour. Look at the module filter in com_modules

Seems so. Module positions are only shown in the dropdown if used. But languages are always shown all possible. So it seems we have that inconsistentcy everywhere, and maybe those other filters need to be fixed?

Anyway, this PR here is fine for me.

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on fb902a4f179fbbc58df44cd14c480d3ee45d3ad5<br /><br />Almost but not quite ;)

The list of components that you can select from does not take into account the menu in use.

Example with two or more menus create a new menu item for the wrapper component in just one of the menus

Now check the component filter in that menu and you see the wrapper component - all good

Now check the component filter in the other menu. I expect not to see the wrapper component being listed but it is  ;(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

pe7er committed:
Added currentMenuType context to the component filter

pe7er commented:
Thanks @brianteeman 
I've added the currentMenuType context to the component filter.
Now the - Select Component - filter only shows the component views that are used in that specific menu.

![joomla-menu](https://user-images.githubusercontent.com/1217850/83141278-adb26900-a0ef-11ea-9774-ac0d156ad0f8.png)

ChristineWk commented:
I have tested this item :white_check_mark: successfully on 585c18556d2bea517ff8f76481ca507b3638a844<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

Quy commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,89 @@
+
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\Field\ListField;
+use Joomla\CMS\Language\Text;
```
```suggestion
use Joomla\CMS\Language\Text;
use Joomla\Database\ParameterType;
```

Quy commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,89 @@
+
+		if ($currentMenuType)
+		{
+			$query->where($db->quoteName('menu.menutype') . ' = ' . $db->quote($currentMenuType));
```
```suggestion
			$query->where($db->quoteName('menu.menutype') . ' = :currentMenuType')
				->bind(':currentMenuType', $currentMenuType);
```

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: Quy <quy@fluxbb.org>

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Thanks!

Co-authored-by: Quy <quy@fluxbb.org>

pe7er commented:
Thanks for rewriting the code using a prepared statement @Quy !

pe7er committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into menu-items-overview-filter-on-component

pe7er committed:
Merge remote-tracking branch 'origin/menu-items-overview-filter-on-component' into menu-items-overview-filter-on-component

pe7er commented:
Could two of you please re-test this PR again? @sere-b @brianteeman @ChristineWk @richard67

ChristineWk commented:
I have tested this item :white_check_mark: successfully on 6e60bd91b80c5c2c5595d384994f4aef24aa3c94<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

pe7er commented:
Thanks @ChristineWk !

richard67 commented:
@pe7er Now the filter offers only those components which are actually used in a menu. But now there is something not nice: When filtering for a component and having selected a menu which contains a menu item for this component, I see a result. When then selecting another menu which doesn't contain that component, I see "No matching results", which is correct. Up to here all fine.

But I don't see that there is a filter applied, because the dropdown value for the component has been cleared, so there is no visual hint showing there is still a filter applied and that I have to use the "Clear" button.

Example with testing sample data: Select the "Fruit Shop" menu, then filter by the "Content" component, then change menu selection to "Australian Parks". One could think that the "Australian Parks" menu is empty when proceeding in this order of processing.

richard67 commented:
@pe7er Sorry, have forgotten to add the question to the end of my previous comment: Is that ok, or would it be better and possible to clear the filter automatically in such case?

richard67 commented:
Maybe it would be the best to clean all filter selections when the selected menu changes?

pe7er commented:
Thank you for your test @richard67 & for spotting that annoying bug.

Yes, I agree that filters that depend on a selected menu, should be cleaned after the selected menu changes.

richard67 commented:
> Yes, I agree that filters that depend on a selected menu, should be cleaned after the selected menu changes.

@pe7er Should that be done with this PR here because this PR makes that annoying thing more visible, or do you wanna make a new separate PR for that?

pe7er commented:
@richard67 I think that I should do it in this PR for the "component view" filter.

Did you notice this annoying behavior after switching selected menu with other filters as well?

richard67 commented:
> Did you notice this annoying behavior after switching selected menu with other filters as well?

No, but that's just because the others don't depend on the selected menu, and so if a filter option is selected and then you change the menu seletion, the selection in the filter is not removed.

In case of the new filter added by this PR, you lose the selection when changing to a menu where the selected value is not available.

If we now clear the filter after a change of the menu selection - which from an UX point makes sense because the menu selection is at the very top, superior to all others, then we can in future PRs pimp the other filters, too, so they only show relevant values for the selected menu (or all) like the new filter does.

pe7er commented:
IMHO we should only clear the filters that are dependent on the menu.

From experience I know that I sometimes use some filters (eg trashed items), then decide to look in other menus. Having to re-select those filters again would be annoying. So I would add the clean filter option only for filters that are dependent on the selected menu.

richard67 commented:
> So I would add the clean filter option only for filters that are dependent on the selected menu.

Sure. Is there any other beside the new filter? I don't remember right now. If we later see that other filters which currently are not dependent on the selected menu should be changed to that, we can do the clear filter for those then in the future PR which changes these filters.

pe7er commented:
I've tried some things to reset the filter when the menu changes, but so far unsuccessfully...

In administrator/components/com_menus/src/Model/ItemsModel.php the method
	protected function populateState($ordering = 'a.lft', $direction = 'asc')
I've changed line 166 to 
```php
// If menutype changed reset pagination and reset componentName filter
if ($menuType != $currentMenuType)
{
	$app->input->set('limitstart', 0);
	$this->setState('filter.componentName', null);
}
else
{
	$componentName = $this->getUserStateFromRequest($this->context . '.filter.componentName', 'componentName');
	$this->setState('filter.componentName', $componentName);
}
```
but that does not work. It keeps receiving the form input with the previous componentName filter.

Can someone please give me a hand with this?

opn365 commented:
I have tested this item :white_check_mark: successfully on 6e60bd91b80c5c2c5595d384994f4aef24aa3c94<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
> I've tried some things to reset the filter when the menu changes, but so far unsuccessfully...
> 
> In administrator/components/com_menus/src/Model/ItemsModel.php the method
> protected function populateState($ordering = 'a.lft', $direction = 'asc')
> I've changed line 166 to
> 
> ```
> // If menutype changed reset pagination and reset componentName filter
> if ($menuType != $currentMenuType)
> {
> 	$app->input->set('limitstart', 0);
> 	$this->setState('filter.componentName', null);
> }
> else
> {
> 	$componentName = $this->getUserStateFromRequest($this->context . '.filter.componentName', 'componentName');
> 	$this->setState('filter.componentName', $componentName);
> }
> ```
> 
> but that does not work. It keeps receiving the form input with the previous componentName filter.
> 
> Can someone please give me a hand with this?

@SharkyKZ Could you have a look on that? Do you have an idea? Thanks in avdance.

rdeutz commented:
I am moving this to 4.1, we are in beta and this is a new feature. Thank you all who have been working on this and bringing it to the state it is now. But if we want to get 4.0 out of the door we need to make a cut.

pe7er commented:
No prob, I understand. Thanks @rdeutz !

PhilETaylor commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_menus
+ *
+ * @copyright   Copyright (C) 2005 - 2020 Open Source Matters, Inc. All rights reserved.
```
```suggestion
 * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

PhilETaylor commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+/**
+ * MenuItem by Component field.
+ *
+ * @since  4.0.0
```
```suggestion
 * @since __DEPLOY_VERSION__
```

PhilETaylor commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+	 * The form field type.
+	 *
+	 * @var     string
+	 * @since   4.0.0
```
```suggestion
	 * @since   __DEPLOY_VERSION__
```

PhilETaylor commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+	 *
+	 * @return    array  An array of JHtml options.
+	 *
+	 * @since   4.0.0
```
```suggestion
	 * @since   __DEPLOY_VERSION__
```

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

pe7er committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: Phil E. Taylor <phil@phil-taylor.com>

pe7er commented:
@PhilETaylor Thanks for your improvements!

Quy commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+
+namespace Joomla\Component\Menus\Administrator\Field;
+
+defined('_JEXEC') or die;
```
```suggestion
\defined('_JEXEC') or die;
```

pe7er replied:
Thanks @Quy !

Quy commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+	{
+		// Initialise variable.
+		$db      = Factory::getDbo();
+		$options = array();
```
```suggestion
		$options = [];
```

pe7er replied:
Thanks @Quy !

fancyFranci commented:
Hi @pe7er, would you like to make this PR ready for 4.2? Then we can finally merge this nice feature :)

ceford commented:
I have tested this item :white_check_mark: successfully on 00b60d949253ff640ad346a3664102d58547e21e<br /><br />Tested in 4.2-dev - also checked only components present in the selected menu are listed in the select drop-down.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

BertaOctech commented:
I have tested this item :white_check_mark: successfully on 00b60d949253ff640ad346a3664102d58547e21e<br /><br />It worked as described. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

alikon committed:
fix update requested

alikon commented:
@pe7er for fix the update requested https://github.com/pe7er/joomla-cms/pull/2

richard67 commented:
> @pe7er for fix the update requested [pe7er#2](https://github.com/pe7er/joomla-cms/pull/2)

I think it will not need new human tests when these changes based on Quy‘s review suggestions will be implemented.

alikon commented:
this pr is in it's second edition  is still pending from ie 2020 2 years ago 
can we speed up the process a little bit ? merge this as is for now ?

richard67 commented:
It needs to be rebased to 4.2-dev, and it needs to apply Quy‘s review suggestions. If @pe7er is ok with it I can do both.

brianteeman commented:
Ideally it also needs to be applied to admin menu items

pe7er committed:
Merge pull request #2 from alikon/patch-2

to fix update requested

pe7er committed:
Merge branch 'joomla:4.1-dev' into menu-items-overview-filter-on-component

pe7er commented:
@richard67 Thanks! 
I updated my PR with the 4.1-dev branch, but I do not know how to rebased to 4.2-dev.
If you would take care of the rebase to 4.2-dev, please do.
Thanks!

brianteeman commented:
It's really easy to rebase

![rebase](https://user-images.githubusercontent.com/1296369/159136063-8470361b-9b14-4cd5-9654-d4bb58cd143d.gif)

richard67 commented:
I've just tried to rebase, but that was not good, it has shown too many changes, so U have reverted that. I think we have to wait with that until the next upmerge from 4.10-dev to 4.2-dev has taken place.

pe7er commented:
@brianteeman Thanks for showing me how to rebase.
@richard67 Thanks! 
If there are any issues with my PR during the "upmerge" then I can redo my PR as a new PR to the 4.2-dev branch.

richard67 commented:
> @brianteeman @richard67 Thanks! If there are any issues with my PR during the "upmerge" then I can redo my PR as a new PR to the 4.2-dev branch.

@pe7er I don't think there will be issues. The problem right now is just that the 4.1-dev branch and also your PR are in advance to the 4.2-dev branch, so the rebase here currently shows all these other changes, too. We just have to wait for the next regular update of the 4.2-dev branch to the 4.1-dev branch (that's the "Upmerge"), and after that the rebase will show the right result.

richard67 committed:
Merge branch '4.1-dev' into menu-items-overview-filter-on-component

richard67 committed:
Merge branch '4.2-dev' into menu-items-overview-filter-on-component

richard67 commented:
Back to pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
I've added back the previous test results in the issue tracker since the change which has invalidated it was just the "\" for the global namespace and short array syntax and after that just a clean banch update and finally a clean rebase.

I'll do a quick test soon and then will set RTC if ok.

brianteeman commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,91 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_menus
+ *
+ * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```

richard67 committed:
Fix copyright year

Co-authored-by: Brian Teeman <brian@teeman.net>

richard67 commented:
I have tested this item :white_check_mark: successfully on 11bca7ec1d25a4c97fc95e836a3a79a7f5742a8b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

pe7er commented:
Thank you all for testing, comments and improvements!

roland-d committed:
Merge branch '4.2-dev' into menu-items-overview-filter-on-component

roland-d commented:
@pe7er Thank you for this addition but I see some unwanted behavior when testing.

## Removed extensions
When I first opened the filters I see this:
![image](https://user-images.githubusercontent.com/359377/159864847-f5090a43-92c0-480d-9883-151065b8fd5b.png)

Notice the empty dropdown. This is what I see opening the dropdown:
![image](https://user-images.githubusercontent.com/359377/159864971-69ca1e12-a24f-4e6b-8c5f-fef04f71804e.png)

The reason this happens is because I have a menu entry for an extension that has been removed after I created the menu item. You get an empty entry because your query links to the `extensions` table that no longer has the name in it. Perhaps we can have a fallback to the `option` value in the menu link?

## Missing translation
In the above screenshot you see `HikaShop`, however when I filter on `Articles` the entry `HikaShop` becomes  `COM_HIKASHOP`. 

![image](https://user-images.githubusercontent.com/359377/159865502-cc87a7da-255c-4e2a-9303-ba072fb851cd.png)

Can you have a look at these 2 issues please? Other than that this looks good. Thank you.

roland-d committed:
Merge branch '4.2-dev' into menu-items-overview-filter-on-component

roland-d commented:
@richard67 In that case you get rid of the empty entry but you cannot filter the menu items then on that particular extension.

richard67 commented:
> @richard67 In that case you get rid of the empty entry but you cannot filter the menu items then on that particular extension.

@roland-d Then add a `->where($db->quoteName('extensions.extension_id') . ' IS NOT NULL')` to the query. Does that help?

pe7er committed:
Merge branch '4.2-dev' of github.com:pe7er/joomla-cms into menu-items-overview-filter-on-component

pe7er committed:
[4.2] Menu Items overview - fix for uninstalled components #28832

pe7er commented:
I've added @richard67 's database improvement to fix the issue with uninstalled components reported by @roland-d 
Thank you both!

pjdevries commented:
Before applying the PR, the `Filter Options` has no `- Select Component -` option, allowing filtration of menu items on component. After applying the PR and following @pe7er's instructions, I was able to filter on various component types, including `com_contact`, showing (just) the `Single Contact` menu item.

After uninstalling a component with an existing menu item, there were no empty entries in the  `- Select Component -` select box. However, the uninstalled component did not show up either, although the menu item still exists. So it appears @roland-d's latest remark has not been addressed (properly) yet.

joomla-bot committed:
Merge tag 'psr12anchor' into psr12/merge/28832

joomla-bot committed:
Phase 1 convert BRANCH to PSR-12

joomla-bot committed:
Merge tag 'psr12final' into psr12/merge/28832

joomla-bot committed:
Phase 2 convert BRANCH to PSR-12

obuisard committed:
Merge branch '4.3-dev' into menu-items-overview-filter-on-component

obuisard committed:
Merge branch '4.3-dev' into menu-items-overview-filter-on-component

obuisard commented:
I have tested this item :white_check_mark: successfully on 40e3f99c0f49848b0115b0908c63f0ae07278725<br /><br />Everything looks good, even after extension uninstall<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

obuisard commented:
I would like to add that when a component is uninstalled, the menu items created with that component become 'orphan' and are no longer 'filtrable'. I would assume this as 'normal' behavior. 

@pjdevries would you mind recording your test in the issue tracker? Thank you.

obuisard commented:
This PR requires updated documentation for the online help such as the one found at https://help.joomla.org/proxy?keyref=Help42:Menus:_Items&lang=en

Thank you!

jwaisner commented:
I have tested this item :white_check_mark: successfully on 40e3f99c0f49848b0115b0908c63f0ae07278725<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

jwaisner commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

obuisard commented:
Great PR, thank you Peter @pe7er for this feature and for all to comment and help make this a good addition. Peter, do you need help with documentation?

heelc29 commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,90 @@
+    protected function getOptions()
+    {
+        // Initialise variable.
+        $db      = Factory::getDbo();
```
```suggestion
        $db      = $this->getDatabase();
```
See https://github.com/joomla/joomla-cms/pull/37097

heelc29 commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,90 @@
+            ->where($db->quoteName('extensions.extension_id') . ' IS NOT NULL');
+
+        $app             = Factory::getApplication();
+        $currentMenuType = $app->input->getString('menutype', $app->getUserState($this->context . '.menutype', ''));
```
```suggestion
        $currentMenuType = $app->getInput()->getString('menutype', $app->getUserState($this->context . '.menutype', ''));
```
See https://github.com/joomla/joomla-cms/pull/39029

heelc29 commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,90 @@
+
+        foreach ($components as $component) {
+            // Load component language files
+            $lang = Factory::getLanguage();
```
```suggestion
            $lang = $app->getLanguage();
```

heelc29 commented on administrator/components/com_menus/src/Field/MenuItemByComponentField.php:
```diff
@@ -0,0 +1,90 @@
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\Field\ListField;
+use Joomla\CMS\Language\Text;
+use Joomla\Database\ParameterType;
```
```suggestion
```
not used

obuisard committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

obuisard committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

obuisard committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

obuisard committed:
Update administrator/components/com_menus/src/Field/MenuItemByComponentField.php

Co-authored-by: heelc29 <66922325+heelc29@users.noreply.github.com>

richard67 commented:
@jwaisner @obuisard Could at least one of you redo your test because there have been some changes? Just to see if it still works, so we can keep RTC. Thanks in advance.

obuisard committed:
Merge branch '4.3-dev' into menu-items-overview-filter-on-component

obuisard commented:
I have tested this item :white_check_mark: successfully on 40e3f99c0f49848b0115b0908c63f0ae07278725<br /><br />2nd testing successful<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28832">issues.joomla.org/tracker/joomla-cms/28832</a>.</sub>

rdeutz committed:
Merge branch '4.3-dev' into menu-items-overview-filter-on-component

obuisard commented:
Hello Peter @pe7er, do you need help with documentation? Would love to sere this PR merged soon.

carlitorweb commented:
Documentation added ✔️ (https://docs.joomla.org/Help4.x:Menus:_Items)

obuisard committed:
Merge branch '4.3-dev' into menu-items-overview-filter-on-component

obuisard commented:
Thank you Peter @pe7er for this PR!
